### PR TITLE
certonly functionality- CSR signing only.

### PR DESCRIPTION
**Pull request by MichelStam**
Hi,

A couple of patches I wrote that add certonly functionality, basically to sign a CSR. Used sometimes for routers, etc.

I have tested the GNUTLS and that works. OpenSSL and MBEDTLS I tested by verifying the workings of the csr_read function, which is where the implementations differ.

Also added a script to dns-01 authentication

### CODE REVIEW

**Comment by ndilieto** [comment]
Hello and thank you for your pull request.

The CSR functionality, can you please explain why you need it and what is your use case?

The command name "certonly" is unfortunate: it does not clearly convey what it does, and it can also be confused with certbot's "certonly" which has different semantics. 

I'm also afraid it goes against the philosophy of uacme, that all certificates are in a well defined configuration directory.

If there is already a certificate the new function only backs it up with a different name, without checking it for validity with the cert_valid function. Putting this in a cron script without care could easily lead to re-issuing unnecessarily and hitting let's encrypt rate limits.

The implementation of the ident array generation with those macros ident_addarg, etc had my head spinning. Why do you need all that complexity? ident is just an array of pointers to strings and the chunk of memory doesn't need to be contiguous.

You are not handling RFC8738 and just ignoring IP identifiers.

The dns-01 script looks reasonable but why is it in the same PR? It looks unrelated to me, please split it in a separate PR.

**Comment by MichelStam** [comment]
Hi ndilieto,

My use case is that I have several routers - Cisco, that need certificates but do not have an ACME implementation. As such I only need signing, the routers provide a CSR.

Certonly can be signonly or something else - its name happens to be the command I used with certbot using ACMEv1 to do the same. Please suggest what would work for you.

I understand the philosophy of uacme, but these routers are not using the same file system, they are different hosts, I don’t see why the certificates have a place there then, can you explain?
Understood on the cron issue, this requires the user to know what they’re doing. I suppose the letsencrypt limitation is to prevent runaway scripts, amongst others.

Ident was something I rigged up to match the **argv array that you use in uacme.c, I needed something compatible without a whole lot of mallocs and stringing those together. I come from an embedded background- since some malloc  implementations allocate in minimum blocks one could waste loads of memory and/or cause fragmentation. This implementation calculates the requirements and implements it using a single buffer. Arguably, it should not have been part of crypto.c.

RFC8738 is not implemented. I wasn’t aware of it and didn’t notice it in uacme either. It could be added, surely, but wouldn’t work with dns-01, which I happen to use for validation. How does uacme treat this?

Basically the script is in the same PR because this adds the functionality that I was looking for in uacme. Feel free to cherry pick it if you like.

Cheers,

Michel Stam

**Comment by ndilieto** [comment]
I understand the router generates the CSR, but I don't understand whether you run uacme on the router or on a computer. Can you please clarify?

**Comment by MichelStam** [comment]
Hello Nikola,

Uacme runs on a local Linux system. The Cisco routers have very limited 
capabilities when it comes to running self-written programs (at least 
the ones I have), C executables are not amongst these.

I will close the pull and push 2 new ones; I fixed a few (small) bugs, 
reworked according to your mail/my answer this morning.

The IP addresses I am looking into, I will have limited ability to test 
it though.

Regards,

Michel

**Comment by MichelStam** [comment]
I will post a reworked version.
